include/thread_pool/worker.hpp

```diff
@@ -226,21 +239,22 @@ inline Worker<Task, Queue>::~Worker()
 template <typename Task, template<typename> class Queue>
 inline void Worker<Task, Queue>::stop()
 {
-    if (m_running_flag.exchange(false, std::memory_order_acq_rel))
+    auto expectedState = State::Running;
+    if (m_state.compare_exchange_strong(expectedState, State::Stopped, std::memory_order_acq_rel))
     {
         wake();
         m_thread.join();
     }
 }
 
 template <typename Task, template<typename> class Queue>
-inline void Worker<Task, Queue>::start(size_t id, WorkerVector& workers, SlottedBag<Queue>& idle_workers, std::atomic<size_t>& num_busy_waiters)
+inline void Worker<Task, Queue>::start(const size_t id, WorkerVector& workers, SlottedBag<Queue>& idle_workers, std::atomic<size_t>& num_busy_waiters)
 {
-    if (m_started_flag.exchange(true, std::memory_order_acq_rel))
-        throw std::runtime_error("The Worker has already been started.");
+    auto expectedState = State::Initialized;
+    if (!m_state.compare_exchange_strong(expectedState, State::Running, std::memory_order_acq_rel))
+        throw std::runtime_error("Cannot start Worker: it has previously been started or stopped.");
 
-    m_running_flag.store(true, std::memory_order_release);
-    m_thread = std::thread(&Worker<Task, Queue>::threadFunc, this, std::ref(id), std::ref(workers), std::ref(idle_workers), std::ref(num_busy_waiters));
+    m_thread = std::thread(&Worker<Task, Queue>::threadFunc, this, id, std::ref(workers), std::ref(idle_workers), std::ref(num_busy_waiters));
```

Comment:
This was the cause of a lot of debugging pain... we were previously passing `id` as an `std::ref`, which caused a dangling reference, and threads would sometimes register with the same id.

Reply:
my fault didn't catch this in the review :)

Reply:
whups !

Reply:
No worries, my fault for writing it :)